File: Graphs/topological sort algorithm/topological_sort.py
from collections import defaultdict

def topological_sort(graph):
    def dfs(node):
        visited.add(node)
        for neighbor in graph[node]:
            if neighbor not in visited:
                dfs(neighbor)
        topo_order.append(node)

    visited = set()
    topo_order = []

    for node in graph:
        if node not in visited:
            dfs(node)

    # The topological order is reversed because we append nodes in reverse order during DFS
    return topo_order[::-1]

# Example usage:
graph = {
    'A': ['B', 'C'],
    'B': ['D'],
    'C': ['D'],
    'D': ['E'],
    'E': []
}

result = topological_sort(graph)
print(result)  # Output: ['A', 'C', 'B', 'D', 'E']

File: Graphs/topological sort algorithm/topological_sort.js
function topologicalSort(graph) {
  const visited = new Set();
  const topoOrder = [];

  function dfs(node) {
    visited.add(node);
    for (const neighbor of graph[node]) {
      if (!visited.has(neighbor)) {
        dfs(neighbor);
      }
    }
    topoOrder.push(node);
  }

  for (const node in graph) {
    if (!visited.has(node)) {
      dfs(node);
    }
  }

  return topoOrder.reverse();
}

// Example usage:
const graph = {
  A: ["B", "C"],
  B: ["D"],
  C: ["D"],
  D: ["E"],
  E: [],
};

const result = topologicalSort(graph);
console.log(result); // Output: ['A', 'C', 'B', 'D', 'E']
